crates/kernel_cmdline/src/utf8.rs

```diff
@@ -0,0 +1,470 @@
+        // later.
+        str::from_utf8(&cmdline)?;
+
+        Ok(Self(bytes::Cmdline::from(cmdline)))
+    }
+
+    /// Returns an iterator over all parameters in the command line.
+    ///
+    /// Properly handles quoted values containing whitespace and splits on
+    /// unquoted whitespace characters. Parameters are parsed as either
+    /// key-only switches or key=value pairs.
+    pub fn iter(&'a self) -> CmdlineIter<'a> {
+        CmdlineIter(self.0.iter())
+    }
+
+    /// Locate a kernel argument with the given key name.
+    ///
+    /// Returns the first parameter matching the given key, or `None` if not found.
+    /// Key comparison treats dashes and underscores as equivalent.
+    pub fn find(&'a self, key: impl AsRef<str>) -> Option<Parameter<'a>> {
+        self.iter().find(|p| p.key() == key.as_ref().into())
+    }
+
+    /// Find all kernel arguments starting with the given UTF-8 prefix.
+    ///
+    /// This is a variant of [`Self::find`].
+    pub fn find_all_starting_with(
+        &'a self,
+        prefix: &'a str,
+    ) -> impl Iterator<Item = Parameter<'a>> + 'a {
+        self.iter().filter(move |p| p.key().starts_with(prefix))
+    }
+
+    /// Locate the value of the kernel argument with the given key name.
+    ///
+    /// Returns the first value matching the given key, or `None` if not found.
+    /// Key comparison treats dashes and underscores as equivalent.
+    pub fn value_of(&'a self, key: impl AsRef<str>) -> Option<&'a str> {
+        self.0.value_of(key.as_ref().as_bytes()).map(|v| {
+            // SAFETY: We know this is valid UTF-8 since we only
+            // construct the underlying `bytes` from valid UTF-8
+            str::from_utf8(v).expect("We only construct the underlying bytes from valid UTF-8")
+        })
+    }
+
+    /// Find the value of the kernel argument with the provided name, which must be present.
+    ///
+    /// Otherwise the same as [`Self::value_of`].
+    pub fn require_value_of(&'a self, key: impl AsRef<str>) -> Result<&'a str> {
+        let key = key.as_ref();
+        self.value_of(key)
+            .ok_or_else(|| anyhow::anyhow!("Failed to find kernel argument '{key}'"))
+    }
+}
+
+/// A single kernel command line parameter key
+///
+/// Handles quoted values and treats dashes and underscores in keys as equivalent.
+#[derive(Clone, Debug, Eq)]
+pub struct ParameterKey<'a>(bytes::ParameterKey<'a>);
+
+impl<'a> std::ops::Deref for ParameterKey<'a> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: We know this is valid UTF-8 since we only
+        // construct the underlying `bytes` from valid UTF-8
+        str::from_utf8(&self.0).expect("We only construct the underlying bytes from valid UTF-8")
+    }
+}
+
+impl<'a> From<&'a str> for ParameterKey<'a> {
+    fn from(input: &'a str) -> Self {
+        Self(bytes::ParameterKey(input.as_bytes()))
+    }
+}
+
+impl<'a> From<bytes::ParameterKey<'a>> for ParameterKey<'a> {
```

Comment:
This is wrong because implementing the trait makes it part of the public api, allowing users to create `utf8::ParameterKey` from a non-UTF-8 `bytes::ParameterKey`.  Should be `TryFrom` instead, and/or moved to private as `from_unchecked`.